Actions/BuildReferenceDocumentation/BuildReferenceDocumentation.HelperFunctions.ps1

```diff
@@ -0,0 +1,294 @@
+        $docfxJson | ConvertTo-Json -Depth 99 | Set-Content -Path $docfxJsonFile -Encoding utf8
+
+        Write-Host "docfx.json:"
+        Get-Content $docfxJsonFile | Out-Host
+
+        # Create new toc.yml
+        Write-Host "Create new toc.yml"
+        $tocYmlFile = Join-Path $docfxpath 'toc.yml'
+
+        Write-Host "Original TOC (from aldoc)):"
+        Get-Content $tocYmlFile | Out-Host
+
+        Set-Content -Path $tocYmlFile -Value ($newTocYml -join "`n") -Encoding utf8
+
+        Write-Host "TOC:"
+        Get-Content $tocYmlFile | Out-Host
+
+        $apps | ForEach-Object {
+            $arguments = @(
+                "build"
+                "--output ""$docfxpath"""
+                "--loglevel $loglevel"
+                "--source ""$_"""
+                )
+            Write-Host "invoke aldoc $arguments"
+            CmdDo -command $aldocPath -arguments $arguments
+        }
+
+        # Set release notes
+        Write-Host "Update index.md"
+        $indexMdFile = Join-Path $docfxpath 'index.md'
+        Set-Content -path $indexMdFile -value $indexContent -encoding utf8
+
+        Write-Host "index.md:"
+        Get-Content $indexMdFile | Out-Host
+
+        $arguments = @(
+            "build"
+            "--output ""$docsPath"""
+            "--logLevel $loglevel"
+            """$docfxJsonFile"""
+            )
+        if ($hostIt) {
+            $arguments += @("-s")
+            Write-Host "Generate and host site"
+        }
+        Write-Host "invoke doxfx $arguments"
+        CmdDo -command docfx -arguments $arguments
+    }
+    finally {
+        Remove-Item -Path $docfxPath -Recurse -Force
+    }
+}
+
+# Build a list of all projects (folders) and apps to use when building reference documentation
+# return value is a hashtable for all apps and a hashtable for all dependencies
+# Every hashtable has project name as key and an array of app files as value
+# if useProjectsAsFolders is false, all apps will be collected in one "project" called "dummy", which is never displayed
+function CalculateProjectsAndApps {
+    Param(
+        [string] $tempFolder,
+        [string[]] $includeProjects,
+        [string[]] $excludeProjects,
+        [switch] $useProjectsAsFolders
+    )
+
+    if ($includeProjects.Count -eq 0) { $includeProjects = @("*") }
+    $projectList = @($includeProjects | ForEach-Object { $_.Replace('\','_').Replace('/','_') })
+    Write-Host "Include Project Patterns"
+    $projectList | ForEach-Object { Write-Host "- $_" }
+    $excludeProjectList = @($excludeProjects | ForEach-Object { $_.Replace('\','_').Replace('/','_') })
+    Write-Host "Exclude Project Patterns"
+    $excludeProjectList | ForEach-Object { Write-Host "- $_" }
+    foreach($mask in 'Apps','Dependencies') {
+        $allApps = @{}
+        foreach($folder in (Get-ChildItem -Path $tempFolder -Filter '*' | Where-Object { $_.PSIsContainer })) {
+            if ($folder.Name -match "^(.*)-main-$mask-(\d*\.\d*\.\d*\.\d*)$") {
```

Comment:
I guess it could be a hotfix from a release branch.

If however people are releasing from a different branch than main or release - it could be problematic if the branch name has a - in it and we could be looking for any branch - then we cannot see what is project and what is branch.